### PR TITLE
Fix unfinished multi-threaded batch analysis and memory leak

### PR DESCRIPTION
**Pull request by uklotzde**
[Conversation on Zulip](https://mixxx.zulipchat.com/#narrow/stream/109171-development/subject/analysis.20performance/near/154270414)

Unfortunately I was not able to trigger the error once again and I'm not able to reason about how this could ever happen. This branch tries to simplify the interaction between the scheduler and its worker threads, reducing the chance for entering an inconsitent state. Maybe the bug is fixed, maybe not.

Includes some extensions that allows to implement [lp1443181](https://bugs.launchpad.net/mixxx/+bug/1443181). I've added a TODO with a reference to this feature request in the code.

### CODE REVIEW

**Comment by uklotzde** [comment]
Includes some extensions that allows to implement [lp1443181](https://bugs.launchpad.net/mixxx/+bug/1443181). I've added a TODO with a reference to this feature request.

**Comment by uklotzde** [comment]
Hopefully also the main memory leak is fixed! I suppose it was the thread-unsafe invocation of QObject::deleteLater(). While analyzing 2100 tracks the memory consumption increased from 1.2 g to 1.3 g. GlobalTrackCache constantly stores only 4 tracks to keep the 4 cores busy.

**Comment by uklotzde** [comment]
I would like to merge this into master asap for broader testing. Until then I have to classify all upcoming bug reports for master as *obsolete*, because this PR represents the status quo of multi-threaded analysis.

File: src/util/compatibility.h
Comment (daschuer) [inline src/util/compatibility.h]: A functor is a class with an operator(). This is only a member function pointer.

File: src/library/dao/cue.cpp
Comment (daschuer) [inline src/library/dao/cue.cpp]: Why do you think that? 
deleteLater() calls QCoreApplication::postEvent() which is thread save.

**Review by daschuer — CHANGES_REQUESTED** [review]
Thank you.

File: src/library/dao/cue.cpp
Comment (uklotzde) [reply src/library/dao/cue.cpp]: Because deleteLater() is a regular slot and the documentation does not explicitly state that it is thread-safe to be called from other threads? I didn't check the source code. Just reasoned according to the rules of the [Qt5 documentation](http://doc.qt.io/qt-5/threads-reentrancy.html):

"**Note**: Qt classes are only documented as thread-safe if they are intended to be used by multiple threads. If a function is not marked as thread-safe or reentrant, it should not be used from different threads. If a class is not marked as thread-safe or reentrant then a specific instance of that class should not be accessed from different threads."

File: src/analyzer/analyzerthread.cpp
Comment (uklotzde) [reply src/analyzer/analyzerthread.cpp]: "eventually" like in "eventually consistent". I will try to revise the comment.

File: src/library/dao/cue.cpp
Comment (daschuer) [reply src/library/dao/cue.cpp]: Ok, so you solution is correct and we can go with it.
However it uses two times the Qt meassage queue.
Did you consider to call QCoreApplication::postEvent() directly?

**Comment by daschuer** [comment]
The missing thread savety hint for deleteLater can be IMHO considered as a short coming of the Qt docs. Otherwise the major use case to use it as smart pointer deleter to get arround race condition at the life time end would be sense less. I would prefere to call deleteLater directly and file a QT doc bug instead. What do you think?

**Comment by uklotzde** [comment]
Done and waiting for approval.

**Comment by daschuer** [comment]
LGTM, Thank you.

**Review by daschuer — APPROVED** [review]

**Pull request merged**
